source/_posts/2017-11-02-kafka-streams-dsl-vs-processor-api.markdown

```diff
@@ -0,0 +1,542 @@
+  () => new PvWindowProcessor("pv-window-store")
+  
+new Topology()
+  (...)
+  .addProcessor("pv-window-processor", pvWindowProcessor, "pv-source")
+```
+
+Now, it's time for event and page view join processor, heart of the topology.
+It seems to be complex but this processor also deduplicates joined stream using `evPvStore`.
+
+``` scala
+class EvJoinProcessor(
+  val pvStoreName: String,
+  val evPvStoreName: String,
+  val joinWindow: FiniteDuration,
+  val deduplicationWindow: FiniteDuration
+) extends AbstractProcessor[ClientKey, Ev] {
+
+  import scala.collection.JavaConverters._
+
+  private lazy val pvStore: WindowStore[ClientKey, Pv] =
+    context().getStateStore(pvStoreName).asInstanceOf[WindowStore[ClientKey, Pv]]
+
+  private lazy val evPvStore: WindowStore[EvPvKey, EvPv] =
+    context().getStateStore(evPvStoreName).asInstanceOf[WindowStore[EvPvKey, EvPv]]
+
+  override def process(key: ClientKey, ev: Ev): Unit = {
+    val timestamp = context().timestamp()
+    val evPvKey = EvPvKey(key.clientId, ev.pvId, ev.evId)
+
+    if (isNotDuplicate(evPvKey, timestamp, deduplicationWindow)) {
+      val pvs = storedPvs(key, timestamp, joinWindow)
+
+      val evPvs = if (pvs.isEmpty) {
+        Seq(EvPv(ev.evId, ev.value, None, None))
+      } else {
+        pvs
+          .filter { pv =>
+            pv.pvId == ev.pvId
+          }
+          .map { pv =>
+            EvPv(ev.evId, ev.value, Some(pv.pvId), Some(pv.value))
+          }
+          .toSeq
+      }
+
+      evPvs.foreach { evPv =>
+        context().forward(evPvKey, evPv)
+        evPvStore.put(evPvKey, evPv)
+      }
+    }
+  }
+
+  private def isNotDuplicate(evPvKey: EvPvKey, timestamp: Long, deduplicationWindow: FiniteDuration) =
+    evPvStore.fetch(evPvKey, timestamp - deduplicationWindow.toMillis, timestamp).asScala.isEmpty
+
+  private def storedPvs(key: ClientKey, timestamp: Long, joinWindow: FiniteDuration) =
+    pvStore.fetch(key, timestamp - joinWindow.toMillis, timestamp).asScala.map(_.value)
+  }
+```
+
+First processor performs a lookup for previously joined `PvEv` by `PvEvKey`.
+If `PvEv` is found the processing is skip because `EvPv` has been already processed.
+
+Next, if page view does not exist for given event (`pvs.isEmpty`) processor returns event itself (left join semantics).
+
+If there are matched page views from given client, 
+try to match page view to event using simple filter `pv.pvId == ev.pvId`.
+We don't need any repartitioning to do that, only gets all page views from given client 
+and join with event in the processor itself. 
+It should be very efficient because every client generates up do hundred page views in 10 minutes.
+
+Perceptive reader noticed that processor also changes the key from `ClientId` to `EvPvKey`
+for deduplication purposes. Without any repartitioning, due to the fact that new key 
+is more detailed that original one. Everything is still within given client context. 
```

Comment:
Everything is still within given client context without the need for any repartitioning. This is possible due to the fact that the new key is more detailed than the original one.